File: Q2.py
import random

def generate_page_reference_string(length):
    return [random.randint(0, 9) for _ in range(length)]

def fifo(page_reference_string, num_frames):
    page_faults = 0
    frame = [-1] * num_frames
    frame_index = 0

    for page in page_reference_string:
        if page not in frame:
            page_faults += 1
            frame[frame_index] = page
            frame_index = (frame_index + 1) % num_frames
        print(f"FIFO: Page {page} -> Frame: {frame}")
    
    return page_faults

def lru(page_reference_string, num_frames):
    page_faults = 0
    frame = [-1] * num_frames
    counter = [0] * num_frames

    for i, page in enumerate(page_reference_string):
        if page not in frame:
            page_faults += 1
            if -1 in frame:
                frame[frame.index(-1)] = page
            else:
                least_recent = counter.index(min(counter))
                frame[least_recent] = page
            counter[frame.index(page)] = i
        else:
            counter[frame.index(page)] = i
        print(f"LRU: Page {page} -> Frame: {frame}")
    
    return page_faults

# Generate a random page-reference string
page_reference_string = generate_page_reference_string(20)

# Apply the string to each algorithm for varying number of page frames
num_frames_range = range(1, 8)
for num_frames in num_frames_range:
    print(f"\nNumber of Frames: {num_frames}")
    print("Page-Reference String:", page_reference_string)
    fifo_page_faults = fifo(page_reference_string, num_frames)
    lru_page_faults = lru(page_reference_string, num_frames)
    print(f"FIFO Page Faults: {fifo_page_faults}")
    print(f"LRU Page Faults: {lru_page_faults}")

File: Q2.c
// Group 62 
// ---------
// Names: Nadege Uwurukundo Mpore ; SI: 300088681
// Names: Noah Ndahirwa;  SI:300152285

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

#define MAX_FRAMES 7
#define PAGE_RANGE 10

// Function to generate a random page-reference string
void generatePageReferences(int pageReferences[], int numReferences) {
    srand(time(NULL));
    for (int i = 0; i < numReferences; i++) {
        pageReferences[i] = rand() % PAGE_RANGE;
    }
}

// Function to check if a page exists in the page frame
bool isPageExists(int pageFrame[], int numFrames, int pageNumber) {
    for (int i = 0; i < numFrames; i++) {
        if (pageFrame[i] == pageNumber) {
            return true;
        }
    }
    return false;
}

// Function to find the index of the page to be replaced using FIFO algorithm
int getFIFOReplacementIndex(int pageFrame[], int numFrames, int currentIndex) {
    return (currentIndex + 1) % numFrames;
}

// Function to find the index of the page to be replaced using LRU algorithm
int getLRUReplacementIndex(int pageFrame[], int numFrames, int pageReferences[], int currentIndex) {
    int lruIndex = 0;
    int lruCounter = currentIndex;
    for (int i = 0; i < numFrames; i++) {
        int j = currentIndex;
        while (j >= 0) {
            if (pageReferences[j] == pageFrame[i]) {
                if (j < lruCounter) {
                    lruCounter = j;
                    lruIndex = i;
                }
                break;
            }
            j--;
        }
        if (j == -1) {
            return i;
        }
    }
    return lruIndex;
}

// Function to simulate the FIFO page-replacement algorithm
int simulateFIFO(int pageReferences[], int numReferences, int numFrames) {
    int pageFaults = 0;
    int currentIndex = 0;
    int pageFrame[MAX_FRAMES] = {0};

    for (int i = 0; i < numReferences; i++) {
        int pageNumber = pageReferences[i];
        if (!isPageExists(pageFrame, numFrames, pageNumber)) {
            pageFaults++;
            int replaceIndex = getFIFOReplacementIndex(pageFrame, numFrames, currentIndex);
            pageFrame[replaceIndex] = pageNumber;
            currentIndex = replaceIndex;
        }
        currentIndex = (currentIndex + 1) % numFrames;
    }
    return pageFaults;
}

// Function to simulate the LRU page-replacement algorithm
int simulateLRU(int pageReferences[], int numReferences, int numFrames) {
    int pageFaults = 0;
    int currentIndex = 0;
    int pageFrame[MAX_FRAMES] = {0};

    for (int i = 0; i < numReferences; i++) {
        int pageNumber = pageReferences[i];
        if (!isPageExists(pageFrame, numFrames, pageNumber)) {
            pageFaults++;
            int replaceIndex = getLRUReplacementIndex(pageFrame, numFrames, pageReferences, currentIndex);
            pageFrame[replaceIndex] = pageNumber;
            currentIndex = replaceIndex;
        }
        currentIndex = (currentIndex + 1) % numFrames;
    }
    return pageFaults;
}

int main() {
    int numReferences = 100;  // Number of page references to generate

    int pageReferences[numReferences];
    generatePageReferences(pageReferences, numReferences);

    printf("Number of page frames\tFIFO Page Faults\tLRU Page Faults\n");
    for (int numFrames = 1; numFrames <= MAX_FRAMES; numFrames++) {
        int fifoPageFaults = simulateFIFO(pageReferences, numReferences, numFrames);
        int lruPageFaults = simulateLRU(pageReferences, numReferences, numFrames);
        printf("%d\t\t\t%d\t\t%d\n", numFrames, fifoPageFaults, lruPageFaults);
    }

    return 0;
}
